From 97050b01bcac2aa398b133535aa558c27021138a Mon Sep 17 00:00:00 2001
From: Hugues Fruchet <hugues.fruchet@foss.st.com>
Date: Tue, 16 Aug 2022 12:04:53 +0200
Subject: [PATCH 0318/1141] media: hantro: reset STM32MP25 VENC before each run

Reset device before each processing to ensure that no
register programming remains from the previous processing.
Workaround for encoder stalled after first encoding.

Signed-off-by: Hugues Fruchet <hugues.fruchet@foss.st.com>
Change-Id: Ideeab1edbb4a64878231a2b34c3a1b34ce3af4a2
---
 drivers/media/platform/verisilicon/hantro_drv.c        | 2 ++
 drivers/media/platform/verisilicon/stm32mp25_venc_hw.c | 3 +++
 2 files changed, 5 insertions(+)

diff --git a/drivers/media/platform/verisilicon/hantro_drv.c b/drivers/media/platform/verisilicon/hantro_drv.c
index d1bb1ed54df5..dfbe1dff6399 100644
--- a/drivers/media/platform/verisilicon/hantro_drv.c
+++ b/drivers/media/platform/verisilicon/hantro_drv.c
@@ -186,6 +186,8 @@ static void device_run(void *priv)
 
 	v4l2_m2m_buf_copy_metadata(src, dst, true);
 
+	ctx->codec_ops->reset(ctx);
+
 	if (ctx->codec_ops->run(ctx))
 		goto err_cancel_job;
 
diff --git a/drivers/media/platform/verisilicon/stm32mp25_venc_hw.c b/drivers/media/platform/verisilicon/stm32mp25_venc_hw.c
index 0aac33afcadc..3dc4ada288a7 100644
--- a/drivers/media/platform/verisilicon/stm32mp25_venc_hw.c
+++ b/drivers/media/platform/verisilicon/stm32mp25_venc_hw.c
@@ -108,6 +108,9 @@ static irqreturn_t stm32mp25_venc_irq(int irq, void *dev_id)
 
 static void stm32mp25_venc_reset(struct hantro_ctx *ctx)
 {
+	struct hantro_dev *vpu = ctx->dev;
+
+	reset_control_reset(vpu->resets);
 }
 
 /*
-- 
2.39.2

